src/Extensions/Subsites/IndexConfigurationExtension.php

```diff
@@ -4,23 +4,63 @@
 
 use SilverStripe\Core\Extension;
 use SilverStripe\SearchService\DataObject\DataObjectDocument;
+use SilverStripe\SearchService\Interfaces\DataObjectDocumentInterface;
 use SilverStripe\SearchService\Interfaces\DocumentInterface;
 
 class IndexConfigurationExtension extends Extension
 {
 
     public function updateIndexesForDocument(DocumentInterface $doc, array &$indexes): void
     {
-        $docSubsiteId = null;
+        // Skip if document object does not implement DataObject interface
+        if (!$doc instanceof DataObjectDocumentInterface) {
+            return;
+        }
+
+        $docSubsiteId = 0;
 
         if ($doc instanceof DataObjectDocument) {
-            $docSubsiteId = $doc->getDataObject()->SubsiteID ?? null;
+            $docSubsiteId = $doc->getDataObject()->SubsiteID ?? 0;
+        }
+
+        $this->updateDocumentWithoutSubsite($doc, $indexes, (int)$docSubsiteId);
+        $this->updateDocumentWithSubsite($doc, $indexes, (int)$docSubsiteId);
+    }
+
+    /**
+     * DataObject does not have a defined SubsiteID. So if the developer explicitly defined the dataObject to be
+     * included in the Subsite Index configuration then allow the dataObject to be added in.
+     */
+    protected function updateDocumentWithoutSubsite(DocumentInterface $doc, array &$indexes, int $docSubsiteId): void
+    {
+        // Document that implement subsite not processed here
+        if ($docSubsiteId) {
+            return;
+        }
+
+        foreach ($indexes as $indexName => $data) {
+            // DataObject explicitly defined on Subsite index definition
+            $explicitClasses = $data['includeClasses'] ?? [];
+
+            if (!isset($explicitClasses[$doc->getDataObject()->ClassName])) {
+                unset($indexes[$indexName]);
+
+                break;
```

Comment:
Won't this mean that if it's in the second index, but not the first, it won't be indexed appropriately?

Reply:
I will test this scenario and update PR.